### ISSUE TITLE
Support for multiple k8s clusters

### ISSUE THREAD

**Issue by 9strands** [Icinga/icinga-kubernetes#95]
So far the initial service looks very interesting and I'm tinkering with it a bit.

However, I think most users are going to wind up needing to monitor multiple clusters - a dev one, a test one, and probably 2 production clusters.  I can't figure out how this can currently work with it - since it needs a specific KUBERNETES_MASTER env var defined to operate.

My guess is that this would also need some changes to the SQL schema, potentially to support with potentially multiple agents running on different nodes polling different clusters - but updating the same backend DB.

**Comment by ngoeddel-openi** [comment]
Oh, I was just browsing through the issues and saw this one.
We also want to have one IcingaWeb and multiple Kubernetes clusters to be included in the UI. It seems there is no way right now to somehow workaround that.
I hope this can be added soon. I really would appreciate that.

**Comment by 9strands** [comment]
I run about 18 K8s clusters, and would love to use -K8sIcinga in order to implement the monitoring (using these: https://github.com/redhat-cop/rhdp-monitoring-scripts/tree/main/openshift/ mostly)

Right now I have custom scripts reaching out from Icinga to each cluster, but having a full-and-proper monitoring solution for each cluster would make my life infinitely easier from a monitoring perspective.

**Comment by PeterLustig1337** [comment]
I also got Issues running icinga-kubernetes to monitor a cluster.
The systemd service is unable to complete start somehow and it always times out.
API request are all being made and I got all cluster information on the web module,
but when it reaches the timeout for service start it just exits.
And everytime this happens the database is somehow corrupt and I need to drop the database and recreate it.
Otherwise i get a lot of "connection reset by peer" errors on the next start.

**Comment by lokidaibel** [comment]
> I also got Issues running icinga-kubernetes to monitor a cluster. The systemd service is unable to complete start somehow and it always times out. API request are all being made and I got all cluster information on the web module, but when it reaches the timeout for service start it just exits. And everytime this happens the database is somehow corrupt and I need to drop the database and recreate it. Otherwise i get a lot of "connection reset by peer" errors on the next start.

Do you try V0.2 Had the same Issues with v0.1.

I would also like multi Cluster Support ! PUSH !

**Comment by sebold-minova** [comment]
Also interested in this

**Comment by lippserd** [comment]
@jhoxhaa implementation for the daemon side should be roughly as follows:

* Create a `cluster` table with a single column `uuid`.
* Include a `cluster_uuid` column in all relevant tables.
* Generate a cluster's `uuid` using the `uid` of the `kube-system` namespace.
* In Icinga DB, we have a similar concept called environments. We utilize `context.WithValue()` [^1] to pass the environment and `ScopedEntity` [^2] to select entities within the current scope. Apply the same approach here.
* Add `ClusterUuid` to `schemav1#Meta` and find a good place where to actually set it.

[^1 ]: https://github.com/Icinga/icingadb/blob/main/cmd/icingadb/main.go#L167
[^2]: https://github.com/Icinga/icingadb/blob/main/pkg/icingadb/sync.go#L90

**Comment by PeterLustig1337** [comment]
> > I also got Issues running icinga-kubernetes to monitor a cluster. The systemd service is unable to complete start somehow and it always times out. API request are all being made and I got all cluster information on the web module, but when it reaches the timeout for service start it just exits. And everytime this happens the database is somehow corrupt and I need to drop the database and recreate it. Otherwise i get a lot of "connection reset by peer" errors on the next start.
> 
> Do you try V0.2 Had the same Issues with v0.1.
> 
> I would also like multi Cluster Support ! PUSH !

yes, got the issue with v0.2

**Comment by 9strands** [comment]
Yes, still waiting for this, and no change in v0.2.  It's new, though.

I noticed the environmentId field in the icingadb function last week (as we've just swung over to icingadb from ido).  I was hoping something similar could be done with Icingadb.
